jetty-maven-plugin/src/main/java/org/eclipse/jetty/maven/plugin/JettyStopMojo.java

```diff
@@ -72,38 +74,132 @@ public void execute() throws MojoExecutionException, MojoFailureException
 
         String command = "forcestop";
 
-        try (Socket s = new Socket(InetAddress.getByName("127.0.0.1"), stopPort);)
+        if (stopWait > 0)
         {
-            OutputStream out = s.getOutputStream();
-            out.write((stopKey + "\r\n" + command + "\r\n").getBytes());
-            out.flush();
-
-            if (stopWait > 0)
+            //try to get the pid of the forked jetty process
+            Long pid = null;
+            try
+            {
+                String response = send(stopKey + "\r\n" + "pid" + "\r\n", true, stopWait);
+                pid = Long.valueOf(response);
+            }
+            catch (NumberFormatException e)
             {
-                s.setSoTimeout(stopWait * 1000);
-                s.getInputStream();
+                getLog().info("Server returned bad pid");
+            }
+            catch (ConnectException e)
+            {
+                //jetty not running, no point continuing
+                getLog().info("Jetty not running!");
+                return;
+            }
+            catch (Exception e)
+            {
+                //jetty running, try to stop it regardless of error
+                getLog().error(e);
+            }
 
-                getLog().info("Waiting " + stopWait + " seconds for jetty to stop");
-                LineNumberReader lin = new LineNumberReader(new InputStreamReader(s.getInputStream()));
-                String response;
-                boolean stopped = false;
-                while (!stopped && ((response = lin.readLine()) != null))
+            //now send the stop command and wait for confirmation - either an ack from jetty, or
+            //that the process has stopped
+            try
+            {
+                if (pid == null)
                 {
+                    //no pid, so just wait until jetty reports itself stopped
+                    getLog().info("Waiting " + stopWait + " seconds for jetty to stop");
+                    String response = send(stopKey + "\r\n" + command + "\r\n", true, stopWait);
+
                     if ("Stopped".equals(response))
-                    {
-                        stopped = true;
                         getLog().info("Server reports itself as stopped");
+                    else
+                        getLog().info("Couldn't verify server as stopped, received " + response);
+                }
+                else
+                {
+                    //wait for pid to stop
+                    getLog().info("Waiting " + stopWait + " seconds for jetty " + pid + " to stop");
+                    send(stopKey + "\r\n" + command + "\r\n", false, 0);
```

Comment:
It's better for `send(...)` to be after `ProcessHandle#of` to avoid potential race condition

Reply:
If the process isn't running when we do the send, then we will just log "Jetty not running".

Reply:
1. Run `send`
2. Execution blocks for unknown reasons
3. Another app starts and system assigns PID previously used by jetty
4. `ProcessHandle` is for unrelated application
5. We get unnecessary stall

Reply:
@XakepSDK good point.

Reply:
Done.